### PR TITLE
feat: Integrate Sign-In Gate Portal into Sticky Bottom Banner

### PR DESCRIPTION
**Pull request by juabara**
## What does this change?

Centralise sign-in gate rendering inside the StickyBottomBanner message picker so banners and the sign‑in gate cannot display at the same time.

Main changes
- Introduce a portal wrapper for sign-in gates:
  - src/components/StickyBottomBanner/SignInGatePortal.tsx — new portal component that uses `useSignInGateSelector` and renders gates into `#sign-in-gate`.
  - Test: src/components/StickyBottomBanner/SignInGatePortal.test.tsx (covers edge cases).
- Make StickyBottomBanner the single source of truth for bottom messages:
  - src/components/StickyBottomBanner.importable.tsx — add sign-in gate as a messagePicker candidate and prioritise message ordering (CMP → SignInGate → Braze → ReaderRevenue).
- Remove legacy gate coordination:
  - src/components/StickyBottomBanner/ReaderRevenueBanner.tsx — remove legacy `signInGateWillShow` dependency from reader revenue canShow logic.
  - Deleted/removed usages of legacy `signInGateWillShow` logic from banner flow.
- Playwright - Banner e2e:
  - Check if auxia/get-treatments endpoint is called on page load.
  - Wrap the sign-in gate selector inside an `Island` in the portal to defer activation until visible.
- Layout updates:
  - Added `host` prop to layouts that require it so StickyBottomBanner can pass host through.

## Why?

- Problem: legacy sign-in gate flow (signInGateWillShow) could not coordinate with the new SDC/Auxia flow and allowed simultaneous display of banners and gates.
- Goal: ensure only one bottom message (banner OR sign-in gate) is visible at a time and keep the system extensible and testable.
- Centralising selection in the StickyBottomBanner messagePicker provides a single priority-based decision point and removes fragile cross-component coordination.

## Testing & validation

- Unit tests: SignInGatePortal unit tests added
- Playwright: Banner e2e was changed to check if auxia/get-treatments is called on page load.

## Notes for reviewers

- I intentionally removed the island & selector from the legacy `withSignInGateSlot` helper; that helper now only injects the `#sign-in-gate` placeholder so StickyBottomBanner owns gate rendering.

### CODE REVIEW

File: dotcom-rendering/src/lib/messagePicker.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { isUndefined, startPerformanceMeasure } from '@guardian/libs';
+import { startPerformanceMeasure } from '@guardian/libs';
```
Comment (tomrf1) [inline dotcom-rendering/src/lib/messagePicker.ts]: Can you explain the changes in this file? Ideally we would not change the messagePicker for this PR

File: dotcom-rendering/src/components/StickyBottomBanner.importable.tsx [hunk 2]
```diff
@@ -65,19 +76,43 @@ const DEFAULT_BANNER_TIMEOUT_MILLIS = 2000;
...
+			}
+		},
 		show: () => {
 			// New CMP is not a react component and is shown outside of react's world
 			// so render nothing if it will show
 			return null;
 		},
 	},
-	timeoutMillis: null,
+	timeoutMillis: 1, // Add 1-millisecond timeout to prevent hanging in problematic environments
```
Comment (tomrf1) [inline dotcom-rendering/src/components/StickyBottomBanner.importable.tsx]: I don't think we want to change this timeout. More generally, are these changes to the CMP banner all to make it work in storybook? It makes me wonder if storybook is the right place to test this logic. We do have Playwright tests ([example](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/playwright/tests/banner.e2e.spec.ts)) which mock services, could this be a better approach?

File: dotcom-rendering/src/lib/messagePicker.ts (hunk 1 above)
Comment (juabara) [reply dotcom-rendering/src/lib/messagePicker.ts]: I changed it because I found a few issues during development:

timeoutify:
- It wasn't cancelling the timeout after recordMessageTimeoutInOphan
- When canShow failed, the catch block only logged the error and never resolved the promise, causing hangs.
- If anything else failed, there was no generic try/catch, also causing hangs.
- Added a finally block to canShow to always cancel the timeout

pickMessage:
- I found an issue with null timeouts + reduce, when promise doesn't resolve or times out, despite of the reduce ensure the logic order, a hanging promise in there was holding the flow indefinetely and it was hard do cancel or guarantee the cleaning of side effects of the already started promises.
- The reduce guarantees the logic order, but doesn't guarantee the effects to happen/complete in order.
- I changed the code to use Promise.allSettled and only then check for the statuses, also I added a timeout to CMP for tests, but having no timeout at all on it may cause the same issue again despite of using Promise.allSettled benefits.
- CMP locally never settles, but regardless the reason, if any promise never resolves or rejects, it will cause the same issue.

If defining a timeout to CMP is not possible, maybe we could think of another approach that is independent of side-effects and respects the priority without blocking a pending canShow.

File: dotcom-rendering/src/components/StickyBottomBanner.importable.tsx (hunk 2 above)
Comment (juabara) [reply dotcom-rendering/src/components/StickyBottomBanner.importable.tsx]: Sorry, I forgot it here during development, CMP never settles locally, but I do recommend defining a timeout here, if CMP fails to resolve/reject the promise for any reason, it will hold everything back as it will never timeout.

File: dotcom-rendering/src/components/StickyBottomBanner.importable.tsx [hunk 3]
```diff
@@ -169,6 +177,46 @@ const buildRRBannerConfigWith = ({
...
+	tags: TagType[],
+	pageId: string,
+	contributionsServiceUrl: string,
+	editionId: EditionId,
+	idUrl: string,
+	host?: string,
+): CandidateConfig<void> => ({
+	candidate: {
+		id: 'sign-in-gate-portal',
+		canShow: () =>
```
Comment (tomrf1) [inline dotcom-rendering/src/components/StickyBottomBanner.importable.tsx]: this should trigger the request to /get-treatment

File: dotcom-rendering/src/components/StickyBottomBanner.importable.tsx [hunk 4]
```diff
@@ -169,6 +177,46 @@ const buildRRBannerConfigWith = ({
...
+			canShowSignInGatePortal(
+				isSignedIn,
+				isPaidContent,
+				isPreview,
+				pageId,
+			),
+		show: () => () => (
```
Comment (tomrf1) [inline dotcom-rendering/src/components/StickyBottomBanner.importable.tsx]: this will trigger the call to `createPortal`

File: dotcom-rendering/src/components/StickyBottomBanner.importable.tsx [hunk 5]
```diff
@@ -169,6 +179,50 @@ const buildRRBannerConfigWith = ({
...
+				pageId,
+				contributionsServiceUrl,
+				editionId,
+				contentType,
+				sectionId,
+				tags,
+				retrieveDismissedCount,
+			);
+		},
+		show: (meta: unknown) => () => (
```
Comment (tomrf1) [inline dotcom-rendering/src/components/StickyBottomBanner.importable.tsx]: can this be `meta: AuxiaGateDisplayData`?

File: dotcom-rendering/src/components/StickyBottomBanner/SignInGatePortal.tsx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+		isSignedOut,
+		handleGateDismissed,
+	]);
+
+	if (!(shouldShowGate && targetElement)) {
+		return null;
+	}
+
+	return createPortal(
+		<Island priority="feature" defer={{ until: 'visible' }}>
```
Comment (tomrf1) [inline dotcom-rendering/src/components/StickyBottomBanner/SignInGatePortal.tsx:98]: try removing the island?

File: dotcom-rendering/.storybook/preview.ts [hunk 7]
```diff
@@ -56,7 +56,7 @@ style.appendChild(document.createTextNode(css));
 // Could this be better? Sure, we could investigate ways to really, truly server side
 // render in Storybook but for now this (and some of the other steps we take around
 // hydration) achieve what we need
-window.guardian = {
+(window as any).guardian = {
```
Comment (tomrf1) [inline dotcom-rendering/.storybook/preview.ts:59]: can this be removed?

Comment (juabara) [reply dotcom-rendering/.storybook/preview.ts:59]: When I remove this, it gives an error

Property 'guardian' does not exist on type 'Window & typeof globalThis'.ts(2339)

Comment (tomrf1) [reply dotcom-rendering/.storybook/preview.ts:59]: I just checked and this is also the case on main

File: dotcom-rendering/src/components/StickyBottomBanner/SignInGatePortal.tsx (hunk 6 above)
Comment (juabara) [reply dotcom-rendering/src/components/StickyBottomBanner/SignInGatePortal.tsx:98]: It was decide to keep it so it only appears when the user scrolls down the page.

File: dotcom-rendering/src/components/StickyBottomBanner/SignInGatePortal.tsx [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+	if (
+		!contributionsServiceUrl ||
+		editionId === undefined ||
+		contentType === undefined ||
+		sectionId === undefined ||
+		tags === undefined ||
+		retrieveDismissedCount === undefined
+	) {
+		return Promise.resolve({ show: false });
```
Comment (juabara) [inline dotcom-rendering/src/components/StickyBottomBanner/SignInGatePortal.tsx:153]: Changed it to false, since sign-in gate is not an ab test anymore and now the gate only appears if userTreatment is present.

File: pnpm-lock.yaml [hunk 9]
```diff
@@ -695,7 +695,7 @@ importers:
         version: 0.7.4
       storybook:
         specifier: 8.6.14
-        version: 8.6.14
+        version: 8.6.14(prettier@3.0.3)
```
Comment (tomrf1) [inline pnpm-lock.yaml]: any idea why it's added `(prettier@3.0.3)` all over the place here? The package.json didn't change in this branch

Comment (juabara) [reply pnpm-lock.yaml]: I don't know exactly, reverted the change anyway, the only thing I could find similar to this were dependabot commits.
